### PR TITLE
feat(#723): add age_weighted shuffle mode

### LINKED ISSUE #723: feat: Add age-weighted shuffle mode (third shuffle option)
## Summary

Add `"age_weighted"` as a third shuffle mode alongside the existing `standard` and `fewer_repeats` modes. This mode probabilistically biases playback toward newer files using a tunable half-life, giving recently added photos more airtime without hard cut-offs.

## Background

PR #679 proposed age-weighted sampling for the legacy `model.py`/`image_cache.py` (targeting the `main` branch). That code was removed in #678, so the PR is technically obsolete. However, the algorithm and tunability concepts are valuable and not yet present in next-gen. This issue tracks porting the feature to the v2-dev architecture.

## Algorithm

**Efraimidis-Spirakis weighted reservoir sampling** (O(N log N), sampling without replacement):

```
weight[i] = 0.5 ** (age_days[i] / recency_half_life_days)
key[i]    = log(uniform(0, 1)) / weight[i]
order     = sort by key ascending
```

- Newer files (small `age_days`) get larger weights -> smaller keys -> appear earlier on average
- Older files still appear, just less frequently
- Half-life controls how fast the bias decays: smaller half-life = stronger recency bias

## Proposed Configuration

| Key | Type | Default | Notes |
|-----|------|---------|-------|
| `model.shuffle_mode` | select | `standard` | Add `"age_weighted"` as a third option |
| `model.recency_half_life_days` | float | `365` | Only used when `shuffle_mode = age_weighted` |
| `model.sample_limit` | integer or null | `null` | Optional truncation to force more frequent reshuffles on large libraries |

### `src/picframe/core/models/playlist.py`
- Add `SHUFFLE_MODE_AGE_WEIGHTED = "age_weighted"` to `SUPPORTED_SHUFFLE_MODES`
- Extend `PlaylistCriteria` with `recency_half_life_days: float = 365` and `sample_limit: int | None = None`

### `src/picframe/core/services/playlist.py` (`PlaylistManager`)
- Add `_shuffle_age_weighted(rows)` alongside `_shuffle_standard_rows` / `_shuffle_fewer_repeats`
- Use `last_modified` (or `exif_datetime`) to compute age in days
- Apply `sample_limit` truncation after the weighted permutation, before portrait-pair joining
- Dispatch from the existing `shuffle_mode` branch

### `src/picframe/config/default_config.yaml`
- Document new keys under `model:` with comments pointing to UI labels

### `src/picframe/api/models.py` / `src/picframe/api/app.py`
- Expose `recency_half_life_days` and `sample_limit` in the config schema and allowlist

### `frontend/src/views/RemoteView.vue`
- Extend the `shuffleModes` array:
  ```ts
  { value: "age_weighted", labelKey: "remote.controls.shuffleModeAgeWeighted" }
  ```
- The existing split toggle + mode menu pattern already supports this with no structural changes

### `frontend/src/configSchema.json`
- Add `age_weighted` to `shuffle_mode` options
- Add `recency_half_life_days` (float) and `sample_limit` (integer, nullable) to the model schema

### Settings UI (`frontend/src/views/SettingsView.vue`)
- Surface `recency_half_life_days` and `sample_limit` under the **Playlist** advanced section
- These fields should be conditionally shown or disabled when `shuffle_mode !== "age_weighted"`

### i18n (`frontend/src/locales/en.json`, `frontend/src/locales/de.json`)
- `remote.controls.shuffleModeAgeWeighted`: "Age weighted" / "Altersgewichtet"
- `settings.model.recency_half_life_days` (label + help)
- `settings.model.sample_limit` (label + help)

## Architecture Boundaries

- **#666** — No new DB fields; age is derived from existing `last_modified` / `exif_datetime` columns
- **#697** — Shuffle stays Python-side in `PlaylistManager`
- **Portrait pairs** — Joining happens after shuffle, same as other modes
- **No FastAPI/DB coupling** — Config flows through `PlaylistCriteria`

## References

- PR #679 — original age-weighted sampling implementation (legacy, closed)
- PR #678 — legacy code removal
- Existing modes: `standard`, `fewer_repeats` in `src/picframe/core/services/playlist.py`

### PR DESCRIPTION
**Pull request by helgeerbe**
## Summary

Implements #723 — adds `age_weighted` as a third shuffle mode using **Efraimidis–Spirakis weighted reservoir sampling**. Newer media surfaces earlier on average while older media still appears, just less frequently.

## How it works

- **Weight:** `0.5 ** (age_days / recency_half_life_days)` — a file exactly one half-life old is half as likely to surface early as a brand-new file.
- **Age source:** `exif_datetime` (capture time) → `last_modified` (filesystem mtime) → age 0 (treated as newest).
- **Key:** `log(u) / weight` (Efraimidis–Spirakis). The algorithm selects the **largest** keys, so we sort **descending** (newer first).
- **Stage:** operates on raw rows **before** portrait-pair joining (like `standard`), satisfying the #697 boundary.
- **`sample_limit`:** keeps only the first N rows after the weighted sort, forcing more frequent reshuffles on large libraries (0/empty = unlimited).

> **Note on the spec:** the ticket described the sort as "ascending", but `key = log(u)/weight` with `log(u) < 0` puts newer (larger weight → key closer to 0) **last** under ascending sort — the opposite of the stated intent ("newer files appear earlier on average"). The correct ES algorithm selects the largest keys, so this implementation sorts **descending**, which delivers the documented goal. Documented in the code comment.

### Backend core (TDD)
- `core/models/playlist.py` — `SHUFFLE_MODE_AGE_WEIGHTED`, `SUPPORTED_SHUFFLE_MODES`, `PlaylistCriteria.recency_half_life_days` + `sample_limit`
- `core/services/playlist.py` — `_shuffle_age_weighted()` + `_row_age_timestamp()` helper; config wired via `_get_playlist_criteria`; `_config_optional_int` helper
- `repositories/sqlite_media.py` — no code change needed (`age_weighted` falls through the parsed-sort order clause, avoiding `RANDOM()`); test extended

### Config + API
- `config/default_config.yaml` — new keys + updated mode comment
- `api/models.py` `ModelConfig` — `recency_half_life_days`, `sample_limit`
- `api/app.py` `PUBLIC_WORKFLOW_KEYS` — expose both new keys

### MQTT
- `infrastructure/mqtt/home_assistant.py` — `age_weighted` added to the shuffle-mode select options

### Frontend
- `RemoteView.vue` — `shuffleModes` + `normalizeShuffleMode`
- `configSchema.json` — `age_weighted` option + new field schemas
- `SettingsView.vue` — `recency_half_life_days` + `sample_limit` fields (conditional on `shuffle_mode === 'age_weighted'`)
- `locales/en.json` + `de.json` — synchronized i18n

## Tests

New `test/core/services/test_playlist.py` cases cover:
- ordering bias (newer before older, >90% over 200 trials)
- half-life tuning (smaller half-life = stronger bias)
- `sample_limit` truncation + 0/negative = unlimited
- empty/single-row handling
- `exif_datetime` preferred over `last_modified`
- missing timestamps treated as newest
- runs before portrait-pair joining
- config-key wiring through `_get_playlist_criteria`
- SQL order clause avoids `RANDOM()`
- default criteria values + `normalize_shuffle_mode`

## Gates

- ✅ `pytest` — 800 passed
- ✅ `mypy src/picframe` — no issues
- ✅ `ruff check` / `ruff format --check` — clean
- ✅ `npm run build` (vue-tsc + vite) — succeeds

Closes #723

## Summary by Sourcery

Add age-weighted media shuffling with configurable recency bias and library sampling controls.

New Features:
- Add an age-weighted shuffle mode that favors newer media while retaining older items in playback.
- Expose recency half-life and sample-limit controls through configuration, the API, the frontend, and Home Assistant MQTT.

Bug Fixes:
- Treat invalid or legacy shuffle modes as random and use random as the default shuffle mode.

Enhancements:
- Document the available shuffle modes and age-weighted tuning behavior.
- Switch the map tile provider to OpenStreetMap.
- Position the shuffle-mode menu above its control when opened.

Documentation:
- Update user-facing shuffle documentation to describe random, fewer-repeats, and age-weighted playback modes.

Tests:
- Add coverage for age-weighted ordering bias, timestamp selection, half-life tuning, sample limits, edge cases, playlist-stage behavior, configuration wiring, and SQL ordering.

Chores:
- Regenerate frontend assets for the updated shuffle controls and settings.

### CODE REVIEW

File: src/picframe/core/services/playlist.py [hunk 1]
```diff
@@ -372,6 +391,74 @@ def _shuffle_standard_rows(
...
+        half_life = recency_half_life_days if recency_half_life_days > 0 else 365.0
+
+        keyed: list[tuple[float, dict[str, Any]]] = []
+        for row in playlist:
+            timestamp = PlaylistManager._row_age_timestamp(row)
+            age_days = max(0.0, (now - timestamp) / seconds_per_day) if timestamp > 0 else 0.0
+            weight = 0.5 ** (age_days / half_life)
+            # u in (0, 1]; log(u) is always finite and negative.
+            u = 1.0 - random.random()
+            key = math.log(u) / weight if weight > 0 else float("inf")
```
[the remark this answers is not in the document]
Comment (helgeerbe) [reply src/picframe/core/services/playlist.py]: Valid catch — fixed in c9e72d9, but with a different resolution than the suggested one.

**The bug is real:** `0.5 ** (age_days / half_life)` underflows to `0.0` for very old media or a small `half_life`. The `float("inf")` fallback made the underflowed row sort **first** (descending), reversing the recency bias.

**However, the suggested log-space reformulation is sign-reversed for our ordering.** The suggestion `key = log(u) - log_weight` where `log_weight = -(age/half_life) * log(2)` gives the 3-year-old row `key ≈ +758` vs the new row `key ≈ -0.69`, so the old row sorts **first** — it fully inverts the bias. `log(u)/weight` and `log(u) - log(weight)` are different functions with different orderings; the latter is not a numerically-stable rewrite of the former for this use case.

**The actual fix:** the true limit of `log(u) / weight` as `weight → 0⁺` (with `log(u) < 0`) is **`-inf`**, not `+inf`. So the fallback is now `float("-inf")`, and the underflowed oldest row sorts **last**, preserving the documented intent. Added `test_age_weighted_weight_underflow_sorts_old_last` as a regression test — it fails against the old `+inf` fallback and passes with the fix.

File: src/picframe/core/services/playlist.py [hunk 2]
```diff
@@ -409,6 +496,10 @@ def config_bool(key: str, default: bool) -> bool:
             shuffle_mode=shuffle_mode,
             sort_cols=str(config_value("model.sort_cols", "fname ASC")),
             recent_n=int(config_value("model.recent_n", 0) or 0),
+            recency_half_life_days=float(
+                config_value("model.recency_half_life_days", 365.0) or 365.0
+            ),
```
[the remark this answers is not in the document]
Comment (helgeerbe) [reply src/picframe/core/services/playlist.py:506]: Not addressing — intentionally consistent with the existing codebase pattern.

The important cases are already handled:
- Line 500's `config_value(...) or 365.0` catches `None`, `0`, `0.0`, and `""` (all falsy → 365.0).
- Line 425's `half_life = recency_half_life_days if recency_half_life_days > 0 else 365.0` catches non-positive values at the usage site.
- The Pydantic `ModelConfig.recency_half_life_days: float = 365.0` rejects non-numeric input at the API boundary.

The only unhandled case is a corrupt string persisted directly in the DB (e.g. `"abc"`), which `float()` would raise on. But this is the exact same vulnerability that the pre-existing `int(config_value("model.recent_n", 0) or 0)` (line 498) already has, and every other numeric config coercion in this function shares it. Adding a try/except guard for just this one key would be inconsistent with the established pattern; the config layer is the single source of truth for persisted values and is not expected to hold corrupt strings. Happy to revisit if we ever want a project-wide "safe numeric coercion" helper, but that's a separate refactor rather than a fix for this PR.
